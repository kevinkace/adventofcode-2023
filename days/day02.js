
export function parseGames(games) {
    return games.split("\n").map((line) => {
        let [
            game,     // "Game 1"
            allRounds // "1 red, 1 green, 1 blue; 1 red, 1 blue; 1 blue"
        ] = line.split(": ");

        return {
            game   : parseInt(game.replace("Game ", ""), 10),

            rounds : allRounds.split("; ").map((round) => {
                return round.split(", ").reduce((moves, /** "1 green" */ move) => {
                    const [ count, color ] = move.split(" ");

                    moves[color.trim()] = parseInt(count, 10);

                    return moves;
                }, {});
            })
        };
    });
}