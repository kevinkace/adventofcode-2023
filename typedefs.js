/**
 * @typedef {Object} Game
 * @property {number} game
 * @property {Round[]} rounds
 */

/**
 * @typedef {Object} Round
 * @property {number} round
 * @property {number} red
 * @property {number} green
 * @property {number} blue
 */

export default {};
