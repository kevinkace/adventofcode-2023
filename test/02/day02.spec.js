import { readFile, writeFile } from "node:fs/promises";
import { fileURLToPath }       from "node:url";
import path                    from "node:path";

import { expect, test, describe } from "vitest";

import { parseGames, validateGame, sumValidateGames, fewestGameCubes, powerCubes } from "../../days/day02.js";

import typedefs from "../../typedefs.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);

const input    = await readFile(path.join(__dirname, "input.txt"), "utf8");
const sample01 = await readFile(path.join(__dirname, "sample01.txt"), "utf8");


/** @type  */
const availCubes = {
    red   : 12,
    green : 13,
    blue  : 14
};

describe("day 02", () => {
    test("can parse input", () => {
        expect(parseGames(sample01)).toMatchSnapshot();
    });

    test("can validate a game", () => {
        const parsed = parseGames(sample01);
        const results = [ true, true, false, false, true ];

        parsed.forEach((game, idx) => {
            expect(validateGame(game, availCubes)).toBe(results[idx]);
        });
    });

    test("can sum valid games from sample", () => {
        const parsed = parseGames(sample01);

        expect(sumValidateGames(parsed, availCubes)).toBe(8);
    });

    test("can sum valid games", () => {
        const parsed = parseGames(input);
        const result = sumValidateGames(parsed, availCubes);

        console.log(result);

        expect(result).toBeTypeOf("number");
    });

    test("can get fewest cubes for 1 game", () => {
        const parsed = parseGames(sample01);
        const result = fewestGameCubes(parsed[0]);

        expect(result).toMatchSnapshot();
    });

    test("can get fewest cubes for all games", () => {
        const parsed = parseGames(sample01);
        const result = parsed.map(fewestGameCubes);

        expect(result).toMatchSnapshot();
    });

    test('can power cubes', () => {
        const parsed = parseGames(sample01);
        const result = parsed
            .map(fewestGameCubes)
            .map(powerCubes);

        console.log(result);

        expect(result).toMatchSnapshot();
    });

    test('can sum power cubes', () => {
        const parsed = parseGames(sample01);
        const result = parsed
            .map(fewestGameCubes)
            .map(powerCubes)
            .reduce((sum, cubes) => {
                return sum + cubes;
            }, 0);

        expect(result).toMatchSnapshot();
    });
});
