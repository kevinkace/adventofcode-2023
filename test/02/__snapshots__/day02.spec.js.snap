// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`day 02 > can get fewest cubes for 1 game 1`] = `
{
  "blue": 6,
  "green": 2,
  "red": 4,
}
`;

exports[`day 02 > can get fewest cubes for all games 1`] = `
[
  {
    "blue": 6,
    "green": 2,
    "red": 4,
  },
  {
    "blue": 4,
    "green": 3,
    "red": 1,
  },
  {
    "blue": 6,
    "green": 13,
    "red": 20,
  },
  {
    "blue": 15,
    "green": 3,
    "red": 14,
  },
  {
    "blue": 2,
    "green": 3,
    "red": 6,
  },
]
`;

exports[`day 02 > can parse input 1`] = `
[
  {
    "game": 1,
    "rounds": [
      {
        "blue": 3,
        "red": 4,
      },
      {
        "blue": 6,
        "green": 2,
        "red": 1,
      },
      {
        "green": 2,
      },
    ],
  },
  {
    "game": 2,
    "rounds": [
      {
        "blue": 1,
        "green": 2,
      },
      {
        "blue": 4,
        "green": 3,
        "red": 1,
      },
      {
        "blue": 1,
        "green": 1,
      },
    ],
  },
  {
    "game": 3,
    "rounds": [
      {
        "blue": 6,
        "green": 8,
        "red": 20,
      },
      {
        "blue": 5,
        "green": 13,
        "red": 4,
      },
      {
        "green": 5,
        "red": 1,
      },
    ],
  },
  {
    "game": 4,
    "rounds": [
      {
        "blue": 6,
        "green": 1,
        "red": 3,
      },
      {
        "green": 3,
        "red": 6,
      },
      {
        "blue": 15,
        "green": 3,
        "red": 14,
      },
    ],
  },
  {
    "game": 5,
    "rounds": [
      {
        "blue": 1,
        "green": 3,
        "red": 6,
      },
      {
        "blue": 2,
        "green": 2,
        "red": 1,
      },
    ],
  },
]
`;

exports[`day 02 > can power cubes 1`] = `
[
  48,
  12,
  1560,
  630,
  36,
]
`;

exports[`day 02 > can sum power cubes 1`] = `2286`;
